### ISSUE TITLE
dropout layer performance

### ISSUE THREAD

**Issue by r2007** [Lasagne/Lasagne#202]
original:
```
    def get_output_for(self, input, deterministic=False, **kwargs):
        if deterministic or self.p == 0:
            return input
        else:
            retain_prob = 1 - self.p
            if self.rescale:
                input /= retain_prob

            # use nonsymbolic shape for dropout mask if possible
            input_shape = self.input_shape
            if any(s is None for s in input_shape):
                input_shape = input.shape

            return input * _srng.binomial(input_shape, p=retain_prob,
                                          dtype=theano.config.floatX)
```
suggestion(draft not test)
```
    def get_output_for(self, input, deterministic=False, **kwargs):
        if self.p == 0:
            return input
        retain_prob = 1 - self.p
        if deterministic:
            if self.rescale:
                input *= retain_prob
            return input
        else:
            # use nonsymbolic shape for dropout mask if possible
            input_shape = self.input_shape
            if any(s is None for s in input_shape):
                input_shape = input.shape

            return input * _srng.binomial(input_shape, p=retain_prob,
                                          dtype=theano.config.floatX)
```
Reason: remove scale step from training process. If the prediction speed is also important, we can setup a custom network using retain_prob * (original next layer W) as the new next layer weight,
Sorry for my native language is not English

**Comment by benanne** [comment]
I'm not a fan of this personally, it causes interactions between weight initialization, learning rates, dropout probabilities etc. to be much more complicated, so parameter tuning becomes harder. We have  also discussed this in the past and decided to scale at train time (see #10).

If this is significantly affecting performance we could reconsider, but we'd need to see some (reproducible) benchmarks first :)

**Comment by r2007** [comment]
just a brief test(MAC bookair wo GPU), it seems no effect.

original mnist.py
```
Starting training...
Epoch 1 of 500 took 14.701s
  training loss:		1.360253
  validation loss:		0.474494
  validation accuracy:		87.14 %%
Epoch 2 of 500 took 14.967s
  training loss:		0.599835
  validation loss:		0.335698
  validation accuracy:		90.47 %%
Epoch 3 of 500 took 14.926s
  training loss:		0.472754
  validation loss:		0.283821
  validation accuracy:		91.80 %%
Epoch 4 of 500 took 14.706s
  training loss:		0.404384
  validation loss:		0.252648
  validation accuracy:		92.52 %%
Epoch 5 of 500 took 14.897s
  training loss:		0.366356
  validation loss:		0.229883
  validation accuracy:		93.15 %%
Epoch 6 of 500 took 14.699s
  training loss:		0.335713
  validation loss:		0.211718
  validation accuracy:		93.68 %%
Epoch 7 of 500 took 14.705s
  training loss:		0.311655
  validation loss:		0.196767
  validation accuracy:		94.21 %%
Epoch 8 of 500 took 14.650s
  training loss:		0.288381
  validation loss:		0.184789
  validation accuracy:		94.52 %%
Epoch 9 of 500 took 14.773s
  training loss:		0.268890
  validation loss:		0.172883
  validation accuracy:		94.92 %%
Epoch 10 of 500 took 14.760s
  training loss:		0.260596
  validation loss:		0.164135
  validation accuracy:		95.19 %%
```
mnist.py with rescale=False dropout layers
```
Starting training...
Epoch 1 of 500 took 14.641s
  training loss:		2.174004
  validation loss:		1.218519
  validation accuracy:		70.08 %%
Epoch 2 of 500 took 14.832s
  training loss:		1.512591
  validation loss:		0.638280
  validation accuracy:		80.85 %%
Epoch 3 of 500 took 15.066s
  training loss:		0.904060
  validation loss:		0.634179
  validation accuracy:		85.59 %%
Epoch 4 of 500 took 14.903s
  training loss:		0.691725
  validation loss:		0.640491
  validation accuracy:		87.57 %%
Epoch 5 of 500 took 14.595s
  training loss:		0.590939
  validation loss:		0.644233
  validation accuracy:		88.62 %%
Epoch 6 of 500 took 14.653s
  training loss:		0.530689
  validation loss:		0.625714
  validation accuracy:		89.60 %%
Epoch 7 of 500 took 14.719s
  training loss:		0.492212
  validation loss:		0.618717
  validation accuracy:		90.06 %%
Epoch 8 of 500 took 14.875s
  training loss:		0.463414
  validation loss:		0.605053
  validation accuracy:		90.48 %%
Epoch 9 of 500 took 14.718s
  training loss:		0.432888
  validation loss:		0.592955
  validation accuracy:		90.74 %%
Epoch 10 of 500 took 14.796s
  training loss:		0.415522
  validation loss:		0.576253
  validation accuracy:		91.05 %%
```

**Comment by benanne** [comment]
Okay. It would be nice to be able to reproduce this on other machines to see if it maybe has a noticeable effect for some use cases / hardware setups. But my gut feeling is that it won't. I will close this issue, but if anyone observes a noticeable performance improvement from changing this behaviour, feel free to reopen it!
